tools/src/main/scala/scala/scalanative/build/NativeLib.scala

```diff
@@ -1,27 +1,95 @@
 package scala.scalanative
 package build
 
+import java.io.File
 import java.nio.file.Path
+import java.util.regex._
 
 /** Original jar or dir path and generated dir path for native code */
 private[scalanative] case class NativeLib(src: Path, dest: Path)
 
+/** Utilities for dealing with native library code */
 private[scalanative] object NativeLib {
+  private val fs           = File.separator
+  private val jarExtension = ".jar"
 
-  /** List of source patterns used */
-  val jarExtension  = ".jar"
-  val srcExtensions = Seq(".c", ".cpp", ".S", ".h", ".hpp")
-  val srcPatterns   = srcExtensions.mkString("glob:**{", ",", "}")
+  /** Object file extension: ".o" */
+  val oExt = ".o"
+
+  /** C++ file extension: ".cpp" */
+  val cppExt = ".cpp"
+
+  /** List of source patterns used: ".c, .cpp, .S" */
+  val srcExtensions = Seq(".c", cppExt, ".S")
+
+  /**
+   * Name of directory that contains native code: "scala-native"
+   */
+  val codeDir = "scala-native"
+
+  /** Used to find native source files in directories */
+  val srcPatterns = srcExtensions.mkString(
+    s"glob:**${fs}classes${fs}${codeDir}${fs}**{",
+    ",",
+    "}")
+
+  /** Used to find native source files in jar files */
+  private val jarSrcRegex: String = {
+    val regexExtensions = srcExtensions.mkString("""(\""", """|\""", ")")
+    s"""^${codeDir}${fs}(.+)${regexExtensions}$$"""
+  }
+
+  /**
+   * This method guarantees that only code copied and generated
+   * into the `native` directory and also in the `scala-native`
+   * sub directory gets picked up for compilation.
+   *
+   * @param workdir the base working directory
+   * @return the source pattern
+   */
+  def destSrcPatterns(workdir: Path): String = {
+    val workdirName = workdir.getFileName()
+    srcExtensions.mkString(
+      s"glob:**${fs}${workdirName}${fs}*${fs}${codeDir}${fs}**{",
```

Comment:
This pattern looks weird. Why does it only take the file name part of `workdir` into account? It should use the full path. Or it shouldn't even be part of the patterns, but the search should be restricted to be inside `workdir`.

Reply:
Yes, I think I could do that and remove the `**` from the front of the `glob`. I will try that.

Reply:
This is fixed too to use full path and only the unpacked dirs which is basically the correct solution.